core/cache/controller_test.go

```diff
@@ -224,17 +213,51 @@ func (s *ControllerSuite) TestRemoveUnit(c *gc.C) {
 	s.AssertResident(c, unit.CacheId(), false)
 }
 
+func (s *ControllerSuite) TestMarkAndSweep(c *gc.C) {
+	controller, events := s.new(c)
+
+	// Note that the model change is processed last.
+	s.processChange(c, charmChange, events)
+	s.processChange(c, appChange, events)
+	s.processChange(c, machineChange, events)
+	s.processChange(c, unitChange, events)
+	s.processChange(c, modelChange, events)
+
+	done := make(chan struct{})
+	go func() {
+		// Removals are congruent with FILO.
+		// Model is last because models are added if they do not exist,
+		// when we first get a delta for one of their entities.
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveUnit{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveMachine{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveApplication{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveCharm{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveModel{})
+		done <- struct{}{}
+	}()
+
+	c.Assert(controller.Marked(), jc.IsFalse)
+	controller.Mark()
+	c.Assert(controller.Marked(), jc.IsTrue)
+
+	controller.Sweep()
+	<-done
```

Comment:
In this particular case, I happen to know that c.Check won't panic (it queues up a failure to the test, but does not abort the rest of the test).
it would be possible for s.nextChange to panic, though. Or perhaps s.nextChange never returns because we changed the internal logic and so we don't have as many messages.

Better for the test to fail with "timeout after X seconds", than for it to hang until the test runner times it out.